src/main/java/hudson/remoting/Launcher.java

```diff
@@ -456,25 +480,146 @@ private SSLSocketFactory getSSLSocketFactory()
                     cert = certOrAtFilename.getBytes(StandardCharsets.US_ASCII);
                 }
                 try {
-                    keyStore.setCertificateEntry(String.format("alias-%d", i++),
-                            factory.generateCertificate(new ByteArrayInputStream(cert)));
+                    x509Certificates.add((X509Certificate) factory.generateCertificate(new ByteArrayInputStream(cert)));
                 } catch (ClassCastException e) {
                     LOGGER.log(Level.WARNING, "Expected X.509 certificate from " + certOrAtFilename, e);
                 } catch (CertificateException e) {
                     LOGGER.log(Level.WARNING, "Could not parse X.509 certificate from " + certOrAtFilename, e);
                 }
             }
-            // prepare the trust manager
-            TrustManagerFactory trustManagerFactory =
-                    TrustManagerFactory.getInstance(TrustManagerFactory.getDefaultAlgorithm());
-            trustManagerFactory.init(keyStore);
-            // prepare the SSL context
-            SSLContext ctx = SSLContext.getInstance("TLS");
-            ctx.init(null, trustManagerFactory.getTrustManagers(), null);
-            // now we have our custom socket factory
-            sslSocketFactory = ctx.getSocketFactory();
-        }
-        return sslSocketFactory;
+        }
+    }
+
+    private void bootstrapInboundAgent() throws CmdLineException, IOException, InterruptedException {
```

Comment:
By far the most complicated part of this code, this at least is better than the status quo by containing the complicated logic in one method rather than encoding it in the very structure of the classes and entrypoints. Basically, this makes a request to the server's pseudo-JNLP endpoint and based on the result overrides some of the arguments. In the long term I think this mode should cease to exist, which is why I have deprecated it above. But this method should preserve compatibility. Another way this code is better than the status quo is that it reuses more code between this mode and the non `-jnlpUrl` mode.

Reply:
> In the long term I think this mode should cease to exist

Agreed; see https://github.com/jenkinsci/docker-swarm-plugin/pull/125 https://github.com/KostyaSha/yet-another-docker-plugin/issues/288#issuecomment-687212127 etc.